test/common/EnvVars.cpp

```diff
@@ -88,6 +90,117 @@ namespace RcclUnitTesting
+
+          while ((pos = log.find("GPU[", pos)) != std::string::npos) {
+              int gpuIndex = std::stoi(log.substr(pos + 4));
+              std::string::size_type idPos = log.find("Unique ID:", pos);
+              if (idPos == std::string::npos) break;
+              std::string::size_type idEnd = log.find_first_of(" \n", idPos + 11);
+              std::string uniqueId = log.substr(idPos + 11, idEnd - (idPos + 11));
+              uniqueIdToGpuIndexes[uniqueId].push_back(gpuIndex);
+              pos = log.find('\n', pos);
+          }
+
+          // Create a vector of pairs for sorting unique IDs based on the number of associated GPUs
+          std::vector<std::pair<std::string, std::vector<int>>> sortedIds(uniqueIdToGpuIndexes.begin(), uniqueIdToGpuIndexes.end());
+          std::sort(sortedIds.begin(), sortedIds.end(), [](const auto& a, const auto& b) {
+              return a.second.size() > b.second.size();
+          });
+
+          for (const auto& pair : sortedIds) {
+              result.insert(result.end(), pair.second.begin(), pair.second.end());
+          }
+      } catch (const std::exception& e) {
+          std::cerr << "Error: " << e.what() << std::endl;
+          return 1;
+      }
+      if (write(pipefd[1], result.data(), gpuPriorityOrder->size() * sizeof(int)) != gpuPriorityOrder->size() * sizeof(int)) return TEST_FAIL;
+      close(pipefd[0]);
+      close(pipefd[1]);
+      exit(EXIT_SUCCESS);
+    } 
+    else {
+      int status;
+      if (read(pipefd[0], gpuPriorityOrder->data(), gpuPriorityOrder->size() * sizeof(int)) != gpuPriorityOrder->size() * sizeof(int)) return TEST_FAIL;
+      waitpid(pid, &status, 0);
+      assert(!status);
+      close(pipefd[0]);
+      close(pipefd[1]);
+    }
+    return TEST_SUCCESS;
+    return 0;
+  }
+       
+
+  int getDeviceMode (bool *cpxMode){
+    // Prepare parent->child pipe
+    int pipefd[2];
+    if (pipe(pipefd) == -1)
+    {
+      ERROR("Unable to create parent->child pipe for getting the device mode\n");
+      return TEST_FAIL;
+    }
+    pid_t pid = fork();
+    if (0 == pid)
+    {
+      bool iscpxMode = false;
+      try {
+          std::string log = execCommand("rocm-smi --showcomputepartition");
+          bool foundCPX = log.find("CPX") != std::string::npos;
+          if (foundCPX) {
+            iscpxMode = true;
+          }
+      } catch (const std::exception& e) {
+          std::cerr << "Error: " << e.what() << std::endl;
+          return 1;
+      }
+      if (write(pipefd[1], &iscpxMode, sizeof(iscpxMode)) != sizeof(iscpxMode)) return TEST_FAIL;
+      close(pipefd[0]);
+      close(pipefd[1]);
+      exit(EXIT_SUCCESS);
+    }
+    else {
+      int status;
+      if (read(pipefd[0], cpxMode, sizeof(*cpxMode)) != sizeof(*cpxMode)) return TEST_FAIL;
+      waitpid(pid, &status, 0);
+      assert(!status);
+      close(pipefd[0]);
+      close(pipefd[1]);
+    }
+    return TEST_SUCCESS;
+    return 0;
+  }
```

Comment:
I'm curious why `rocm-smi` is run in a fork process?

Reply:
Because we cannot use HIP call prior to launching unless it is inside another child process. You can refer to line 207 for this comment. The same procedure is used for other features such as finding the architecture of the system as in line 211.

Reply:
In your case, I think `popen` already forks and creates a pipe internally, so you may not need your own fork. If so, `getDeviceMode` and `getDevicePriority` can be simplified.